File: src/telemetry/statistics.cpp
#include "telemetry/statistics.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Telemetry
{
Statistics::Statistics(const PbTelemetry::Statistics& protobuf) {
    messages = protobuf.messages();
    averageFrequency = protobuf.averagefrequency();
    seconds = protobuf.seconds();
}

Statistics::operator PbTelemetry::Statistics() const {
    PbTelemetry::Statistics ret;
    ret.set_messages(messages);
    ret.set_averagefrequency(averageFrequency);
    ret.set_seconds(seconds);
    return ret;
}

std::string Statistics::serializeAsJsonString() const {
    PbTelemetry::Statistics protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Statistics::serializeAsProtobufString() const {
    PbTelemetry::Statistics protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Statistics::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::Statistics protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Statistics::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::Statistics protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/telemetry/statistics.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from telemetry import statistics_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class Statistics:
    messages: int = 0
    averageFrequency: int = 0
    seconds: float = 0.0
    
    _proto_message: statistics_pb2.Statistics = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = statistics_pb2.Statistics()

    def _populate_proto(self):
        self._proto_message.messages = self.messages
        self._proto_message.averageFrequency = self.averageFrequency
        self._proto_message.seconds = self.seconds

    @classmethod
    def _from_proto(cls, proto_message) -> "Statistics":
        return cls(
            messages = proto_message.messages,
            averageFrequency = proto_message.averageFrequency,
            seconds = proto_message.seconds,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Statistics":
        message = statistics_pb2.Statistics()
        message.ParseFromString(data)
        return cls(
            messages = message.messages,
            averageFrequency = message.averageFrequency,
            seconds = message.seconds,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Statistics":
        message = statistics_pb2.Statistics()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
